tr/src/lib.rs

```diff
@@ -81,80 +81,75 @@ pub mod runtime_format {
     //!
     //! TODO: better error reporting and support for more replacement option
 
-    /// The result of the runtime_format! macro.
-    /// This implements the Display
-    pub struct FormatArg<'a> {
-        #[doc(hidden)]
-        pub format_str: &'a str,
-        #[doc(hidden)]
-        pub args: &'a [(&'static str, &'a dyn (::std::fmt::Display))],
-    }
-
-    impl<'a> ::std::fmt::Display for FormatArg<'a> {
-        fn fmt(&self, f: &mut ::std::fmt::Formatter<'_>) -> ::std::fmt::Result {
-            let mut arg_idx = 0;
-            let mut pos = 0;
-            while let Some(mut p) = self.format_str[pos..].find(|x| x == '{' || x == '}') {
-                if self.format_str.len() - pos < p + 1 {
-                    break;
-                }
-                p += pos;
-
-                // Skip escaped }
-                if self.format_str.get(p..=p) == Some("}") {
-                    self.format_str[pos..=p].fmt(f)?;
-                    if self.format_str.get(p + 1..=p + 1) == Some("}") {
-                        pos = p + 2;
-                    } else {
-                        // FIXME! this is an error, it should be reported  ('}' must be escaped)
-                        pos = p + 1;
-                    }
-                    continue;
-                }
+    /// Converts the result of the runtime_format! macro into the final String
+    pub fn display_string(format_str: &str, args: &[(&str, &dyn ::std::fmt::Display)]) -> String {
+        use ::std::fmt::Write;
+        let fmt_len = format_str.len();
+        let mut res = String::with_capacity(2 * fmt_len);
```

Comment:
Not sure about the capacity here. It is hard to know what's the require capacity since we don't know the size of the arguments. I would have let the default string growing algorithm do its job. But it's true that 2 times the format length is probably fine.

Reply:
Yeah, there is a trade-off here. I would set at least `fmt_len`, because the final string will almost never be shorter than the format one. If anything `{}` could get an empty string, but that is just two bytes less.

The thing is that if you write `tr!("a {} b {} c {} d")`, you will do 7 _write_ operations, and I guess at least 3 allocations, depending on the length of the arguments.

I've run my benchmarks many times and `2 * fmt_len` seems to be the sweet spot.

And since then I've been reading the `std` library; it uses this handy but undocumented [`Arguments::estimated_capacity`](https://github.com/rust-lang/rust/blob/master/library/core/src/fmt/mod.rs#L375): it does some smart heuristics, and... hey! it doubles some length :nerd_face:!

If you agree, I'd leave it as is.